pkg/workflows/builder.go

```diff
@@ -0,0 +1,73 @@
+package workflows
+
+import (
+	"fmt"
+	"strconv"
+)
+
+// 1. Capability defines JSON schema for inputs and outputs of a capability.
+// Trigger: triggerOutputType := workflowBuilder.addTrigger(DataStreamsTrigger.Config{})
+// Adds metadata to the builder. Returns output type.
+// 2. Consensus: consensusOutputType := workflowBuilder.addConsensus(ConsensusConfig{
+// 	Inputs: triggerOutputType,
+// })
+
+type Workflow struct {
+	spec *workflowSpecYaml
+}
+
+type Trigger[O any] struct {
+	Definition TriggerDefinitionYaml
+	Output     O
+}
+
+type Consensus[O any] struct {
+	Definition StepDefinitionYaml
+	Output     O
+}
+
+type CapabilityDefinition[O any] struct {
+	Ref    string
+	Output O
+}
+
+type NewWorkflowParams struct {
+	Owner string
+	Name  string
+}
+
+func NewWorkflow(
+	params NewWorkflowParams,
+) *Workflow {
+	return &Workflow{
+		spec: &workflowSpecYaml{
+			Owner: params.Owner,
+			Name:  params.Name,
+		},
+	}
+}
+
+func AddTrigger[O any](w *Workflow, trigger Trigger[O]) CapabilityDefinition[O] {
+	// Add ref to trigger.Definition
+	trigger.Definition.Ref = fmt.Sprintf("trigger-%s", strconv.Itoa((len(w.spec.Triggers))))
+	w.spec.Triggers = append(w.spec.Triggers, trigger.Definition)
+
+	return CapabilityDefinition[O]{
+		Output: trigger.Output,
+		Ref:    trigger.Definition.Ref,
+	}
+}
+
+func AddConsensus[O any](w *Workflow, consensus Consensus[O]) CapabilityDefinition[O] {
+	// Add ref to trigger.Definition
```

Comment:
That's also a good reason to pass the capability definition as input instead of the workflow only.